---
title: "MERFISH_HU_1_Preprocessing"
author: "Melissa"
date: "2025-01-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(ggExtra)
library(cowplot)
library(ggplot2)
library(Seurat)
library(sctransform)
library(dplyr)
library(viridis)
library(SpatialExperiment)
library(SEraster)
library(nnSVG)
library(CooccurrenceAffinity)

library(MERINGUE)
library(Seurat)
library(dplyr)

```

```{r}
gtheme= theme (axis.text.x = element_text(size=12, face="bold", colour = "black", angle = 0, hjust = 0.5),axis.text.y = element_text(size=12, face="bold", colour = "black"),legend.text = element_text(face = "bold", size = 12),text=element_text(size=14, face="bold", colour = "black"))
###
themplot=theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),axis.ticks.x=element_blank())

'%!in%' <- function(x,y)!('%in%'(x,y))
```


```{r}
'%!in%' <- function(x,y)!('%in%'(x,y))
mainDir=here::here() #C:/Users/Melissa S/Desktop/PROJECTS/Spatial_Transcriptomics/MERFISH
inputDir= file.path(mainDir,"input")
outputDir= file.path(mainDir,"output")

RDatadir <- file.path(outputDir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}
figDir= file.path(outputDir, "Figures") ; if(!file.exists(figDir)){dir.create(figDir)}


samples=list.files(inputDir)
col_samples= rev(c(  "#F0E492", "#FFB6C1","#0072B2", "#D55E00", "#CC79A7", "#006400","#eb9d01", "#0b1b7f", "#ff9f7d", "#56B4E9", "#7F658A", "#2B6968", "#A8A1E3")) #"#32CD32",
names(col_samples)=paste0("seu_",samples)
```

#-__-_----------------------- STEP1: LOAD & QC the SAMPLES -----------------------####

Read all the datasets and merge them:
```{r}

## read in data
for(i in seq(1, length(samples))){

  cd <- read.csv(file.path(inputDir, samples[i],"cell_by_gene.csv"))
  annot <- read.csv(file.path(inputDir, samples[i],'cell_metadata.csv'))

  annot$cell=paste0(samples[i], "_cell", seq(1, nrow(annot)))

  rownames(annot)=annot$cell
  rownames(cd)=annot$cell


  good.genes <- colnames(cd)[!grepl('Blank', colnames(cd))]
  counts <- t(cd[, good.genes])
  counts=counts[-1,]
  colnames(counts)=paste0(samples[i],"_cell", seq(1, ncol(counts)))
  #counts=MERINGUE::cleanCounts(counts, verbose = FALSE, min.lib.size = 5, min.reads =5)
  annot= annot %>% dplyr::filter(cell %in% colnames(counts))
  stopifnot(identical(colnames(counts), rownames(annot)))


assign(paste0("seu_", samples[i]), CreateSeuratObject(counts=counts, meta.data = annot, project=samples[i], min.cells = 5, min.features = 5))

}
```

Perform a global QC analysis on all the samples, exclude samples with too few or too many cells:
```{r}
seu= mget(paste0("seu_", samples))

for(i in seq_len(length(seu))){
  seu[[i]]$sample= names(seu)[i]
  
  annot <- read.csv(file.path(inputDir, samples[i],'cell_metadata.csv'))

  pos <- annot[, c('center_x', 'center_y')]
  pos <- annot[, c('center_x', 'center_y')]
  colnames(pos)=c("center_1", "center_2")
  pos=as.matrix(pos)
  pos[,2] <- -pos[,2]
  rownames(pos)= paste0(samples[i], "_cell", seq(1, nrow(pos)))
  pos= pos[which(rownames(pos) %in% colnames(seu[[i]])),]
  stopifnot(identical(rownames(pos), colnames(seu[[i]])))
  
  
  seu[[i]][['spatial']] <- CreateDimReducObject(embeddings = pos, key = 'center_', assay = 'RNA')
  rm(pos)
  rm(annot)

}

qs::qsave(seu, file.path(RDatadir, "SeuratList_WithSpatial_Raw.qs"))
seur=qs::qread(file.path(RDatadir, "SeuratList_WithSpatial_Raw.qs"))


##Filter out the low quality cells:
so=merge(seur[[1]], seur[-1])
#######
dt=as.data.frame(table(so$sample))
rownames(dt)= dt$Var1
dt=dt[v1$sample,]

v1=aggregate(so$nCount_RNA, by=list(so$sample), mean)
v2=aggregate(so$nFeature_RNA, by=list(so$sample), mean)
v1$nFeature_RNA= v2$x
colnames(v1)=c("sample", "nCount_RNA", "nFeature_RNA")
v1$Nbr_cell_Raw= dt$Freq
writexl::write_xlsx(v1, file.path(figDir, "RawNumbers_Mean_nCount_nFeature.xlsx"))

#so= subset(so, nCount_RNA > 50 & nFeature_RNA > 15)

so= subset(so, nCount_RNA > 20 & nFeature_RNA > 10)

dt=as.data.frame(table(so$sample))
rownames(dt)= dt$Var1
dt=dt[v1$sample,]
v1=aggregate(so$nCount_RNA, by=list(so$sample), mean)
v2=aggregate(so$nFeature_RNA, by=list(so$sample), mean)
v1$nFeature_RNA= v2$x
colnames(v1)=c("sample", "nCount_RNA", "nFeature_RNA")
v1$Nbr_cell_Raw= dt$Freq
writexl::write_xlsx(v1, file.path(figDir, "Filtered_Numbers_Mean_nCount_nFeature.xlsx"))


so=subset(so, sample != "seu_region1_463172") #Too many cells
so=subset(so, sample !="seu_region0_AG0377") #Too few cells < 3000 

so=subset(so, sample !="seu_region1_AW6196") #Too few cells < 3000

so=subset(so, sample !="seu_region0_AM4078") #
so=subset(so, sample !="seu_region1_AG0377")

so=subset(so, sample !="seu_region1_AL9879")
so=subset(so, sample !="seu_region0_AL9879")
```

```{r}
#First try logCPM normalization:

so= so %>%  JoinLayers() %>% NormalizeData( normalization.method = "LogNormalize", scale.factor = 10000)  %>% ScaleData() %>% FindVariableFeatures( selection.method = "vst", nfeatures = 200) %>% RunPCA( npcs = 30)   %>% RunUMAP(., reduction= "pca", dims= 1:20 ) %>% FindNeighbors(., reduction = "pca", dims = 1:20) %>% FindClusters(., resolution = 0.8)
```


Find Top Markers:
```{r}
so= so %>% FindClusters(., resolution = 0.8)
####

###
Idents(so)="seurat_clusters"
topMarkers= FindAllMarkers(so, logfc.threshold = 0.4, only.pos = TRUE) %>% dplyr::filter(p_val_adj < 0.05)


writexl::write_xlsx(topMarkers, paste0(RDatadir, "/DEG_1vsAll.xlsx"))
```

Annotate cell clusters:
```{r}
so$type=so$seurat_clusters

so$type=dplyr::recode(so$type, "9"="Endothelial", "7"="Fibroblast", "2"="Basal", "8"="Adipocyte", "6"="LP2_b", "4"="Immune", "10"="HS", "1"="Fibroblast", "0"="Fibroblast3",  "3"="Endothelial", "12"="Fibroblast2", "5"="LP", "11"="LP2_a", "13"="Fibroblast2", "14"="Fibroblast3")
###-----------------------------#####
Idents(so)="type"
topMarkers= FindAllMarkers(so, logfc.threshold = 0.4, only.pos = TRUE) %>% dplyr::filter(p_val_adj < 0.05)
writexl::write_xlsx(topMarkers, paste0(RDatadir, "/DEG_1vsAll_Type_Final.xlsx"))
```


Barplot of the distribution :
```{r}
pdf(file.path(figDir, "BARPLOT_Sample_Type.pdf"))
dt <- as.data.frame(table(so$type,so$sample))

p <- ggplot(dt ,aes(x= Var2, y= Freq, fill= Var1))+ 
  geom_bar(stat="identity", position = "fill",show.legend = TRUE,col="black",width=0.8) + scale_fill_manual(values=col_type) + themplot + gtheme + theme (axis.text.x = element_text(size=12, face="bold", colour = "black", angle = 45, hjust = 1)) + xlab("") + ylab("")
print(p)
dev.off()
```

UMAP:
```{r}
cells.order.random <- sample(Cells(so))
dt= Embeddings(so, reduction = "umap")[cells.order.random,]
dt= cbind(dt, so@meta.data[cells.order.random,])
#####
g= dt %>% ggplot(aes(x= umap_1, y=umap_2, fill= type)) +  geom_point( shape = 21, colour="black", stroke=0.7, size=2) + scale_fill_manual(values= col_type) + themplot + NoLegend() + NoAxes()

png(file.path(figDir, "/UMAP_Type.png"), height = 2000, width = 2000, res=300)
print(g)
dev.off()

p= dt %>% ggplot(aes(x= umap_1, y=umap_2, fill= sample)) +  geom_point( shape = 21, colour="black", stroke=0.7, size=2) + scale_fill_manual(values= col_samples) + themplot + NoLegend() + NoAxes()

png(file.path(figDir, "/UMAP_samples.png"), height = 2000, width = 2000, res=300)
print(p)
dev.off()


#get the legend:
g <- dt %>% ggplot(aes(x= umap_1, y=umap_2, fill= type)) +  geom_point( shape = 21, colour="black", stroke=0.7, size=2) + scale_fill_manual(values= col_type) + themplot

g2 <- dt %>% ggplot(aes(x= umap_1, y=umap_2, fill= sample)) +  geom_point( shape = 21, colour="black", stroke=0.7, size=2) + scale_fill_manual(values= col_samples) + themplot


pdf(file.path(figDir, "FinalLegend_Colors.pdf"))
leg=as_ggplot(get_legend(g))
leg2=as_ggplot(get_legend(g2))
print(leg)
print(leg2)
dev.off()

```


Add info about the spatial coordinates:
```{r}
seur=qs::qread(file.path(RDatadir, "SeuratList_WithSpatial_Raw.qs"))

names(seur)[which(names(seur) %!in% so$sample)]


dt= so@meta.data[, c("sample", "type")]
dt$cell=rownames(dt)

for(i in names(seur)){
    print(i)
    dti= dt %>% dplyr::filter(sample== i)
    dti= dti %>% dplyr::filter( cell %in% colnames(seur[[i]]))
    seur[[i]]= subset(seur[[i]], cells=dti$cell)
    
    #dti=dti[match(colnames(seu[[i]]),dti$cell),]
    assertthat::are_equal(dti$cell, colnames(seur[[i]]))
    
    seur[[i]]$type= dti$type
}

qs::qsave(seur, file.path(RDatadir, "New_SeuList_PostProcessed_VF.qs"))

```


PLOT SPATIAL COORDINATES:
```{r}
seur=qs::qread(file.path(RDatadir, "Final_SeuList_PostFiltering_VF.qs"))

for(i in names(seur)){
png(file.path(figDir, paste0("Spatial_2D_",i ,".png")), height = 2000, width = 2000, res = 300)
    ds= Embeddings(seur[[i]], reduction = "spatial")
    ds= cbind(ds, seur[[i]]@meta.data)

    g= ds %>% ggplot(aes(x= center_1, y=center_2, fill= type)) +  geom_point( shape = 21, colour="black", stroke=0.7, size=2, alpha=0.7) + scale_fill_manual(values= col_type) + themplot + NoLegend() + NoAxes()
    print(g)
     dev.off()
}
 

for(i in names(seur)){
png(file.path(figDir, paste0("Spatial_2D_V3_",i ,".png")), height = 2000, width = 2000, res = 300)
    ds= Embeddings(seur[[i]], reduction = "spatial")
    ds= cbind(ds, seur[[i]]@meta.data)

    g= ds %>% ggplot(aes(x= center_1, y=center_2, fill= type)) +  geom_point( shape = 21, colour="black", size=2, alpha=0.5 ) + scale_fill_manual(values= col_type) + themplot + NoLegend() + NoAxes()
    print(g)
     dev.off()
} 


for(i in names(seur)){
png(file.path(figDir, paste0("New_col_Spatial_2D_V4_",i ,".png")), height = 2000, width = 2000, res = 300)
    ds= Embeddings(seur[[i]], reduction = "spatial")
    ds= cbind(ds, seur[[i]]@meta.data)

    g= ds %>% ggplot(aes(x= center_1, y=center_2, fill= type)) +  geom_point( shape = 21, colour="black", size=1.5, alpha=0.5 ) + scale_fill_manual(values= col_type) + themplot + NoLegend() + NoAxes()
    print(g)
     dev.off()
}  
```


