---
title: "Step3_6_Hetmap_Alterations"
author: "Melissa"
date: "04/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
mainDir <- here::here() # the MainDir should be the root of Tumor_BRCA_P53:
source(knitr::purl(file.path(mainDir,"Scripts","global_var_current.Rmd"), quiet=TRUE))
source(knitr::purl(file.path(mainDir,"Scripts","scRNAseq_Functions.Rmd"), quiet=TRUE))
source(knitr::purl(file.path(mainDir,"Scripts","global_variables.Rmd"), quiet=TRUE))
```


###############WORK ON THE ST SAMPLE ##############
```{r}
obs <- read.table(paste0(ostcnvDir,"/infercnv.observations.txt")) %>% as.matrix()
colnames(obs) = gsub("[.]1","-1", colnames(obs))
mat <- obs
```

isolate alteration genome for ST:

```{r}
md <- epith@meta.data[which(colnames(epith) %in% colnames(st)), c(14,21,24,25,38)]
###
ccst <- qs::qread(paste0(occDir,"/ConsensusClustering_Result_ST_CNV.qs"))
kp= ccst[[4]]$consensusClass
pc <- as.data.frame(kp)
pc$cell=rownames(pc)
pc= pc %>% column_to_rownames("cell")
all(rownames(pc)== rownames(md))
pc=pc[rownames(md),]
md$k= pc$kp
```

```{r}
pc=pc[rownames(md),]
n <- rownames(pc)
pc=as.data.frame(pc[,1], row.names = n)
colnames(pc)="kp"
```


```{r}
## genes ###
 g=ic_st@gene_order[,1] %>% as.data.frame()
   rownames(g)=rownames(ic_st@gene_order)
  colnames(g)="chromosome"
###
colgene= colorRampPalette(brewer.pal(6, "Dark2"))(length(unique(g$chromosome)))
names(colgene)=unique(g$chromosome)
### cells ####
all(colnames(mat) == colnames(st))

##
mat = mat[,colnames(st)]
all(colnames(mat) == colnames(st))

#### colors ####
 breaks <- c(seq(0.7, 0.99, length.out = 101),seq(1,1.3, length.out = 101))

pc$kp=as.character(pc$kp)
#k=c("#C294C2", "#B16BB3", "#F277F2", "#A65DA0") 
k=c("#CBC2CC", "#B16BB3", "#F277F2", "#A65DA0")
names(k)= as.character(seq_len(4))

wtlcol=list( k, colgene)
names(wtlcol)=c("kp", "chromosome")
```

```{r}
png(file = paste0(hDir, "/HEATMAP_ST_Bright_3.png") ,height=1300,width=1300,res=300)

 pheatmap::pheatmap(mat, show_rownames = F, show_colnames = F,   annotation_col = pc, cluster_rows = F, cluster_cols = T,annotation_row = g, breaks= breaks,color = colorRampPalette(c("navy", "white", "firebrick3"))(202),annotation_colors= wtlcol,clustering_method =  "ward.D",clustering_distance_rows="correlation",
annotation_names_row=T) 
dev.off()
###
png(file = paste0(hDir, "/HEATMAP_ST_Bright_5.png") ,height=1500,width=1500,res=300)

 pheatmap::pheatmap(mat, show_rownames = F, show_colnames = F,   annotation_col = pc, cluster_rows = F, cluster_cols = T,annotation_row = g, breaks= breaks,color = colorRampPalette(c("navy", "white", "firebrick3"))(202),annotation_colors= wtlcol,clustering_method =  "ward.D",clustering_distance_rows="correlation",
annotation_names_row=T) 
dev.off()
```


Do it for P16:

```{r}
obs1 <- read.table(paste0(ostcnvDir,"/infercnv.observations.txt")) %>% as.matrix()
colnames(obs1) = gsub("[.]1","-1", colnames(obs1))
```


```{r}
p16= obs1[,which(colnames(obs1) %in% rownames(cnt@meta.data)[which(cnt$subtype=="P16+ Pre-lesional")]  )]
md <- epith@meta.data[which(colnames(epith) %in% colnames(p16)), c(14,21,24,25,38)]
md=md[colnames(p16),]
#####
ccp <- qs::qread(paste0(occDir,"/CC_Cntr_Cnv_P16/ConsensusClustering_Result_CNV_p16.qs"))
kp= ccp[[4]]$consensusClass
pc <- as.data.frame(kp)
pc$cell=rownames(pc)
all(rownames(pc)== rownames(md))
pc=pc[rownames(md),]
md$k= pc$kp
```


```{r}
n <- rownames(pc)
pc=as.data.frame(pc[,1], row.names = n)
colnames(pc)="kp"
pc$kp=as.character(pc$kp)
```

Part1:
```{r}
# genes ###
 g=ic_cnt@gene_order[,1] %>% as.data.frame()
   rownames(g)=rownames(ic_cnt@gene_order)
  colnames(g)="chromosome"
###
colgene= colorRampPalette(brewer.pal(6, "Dark2"))(length(unique(g$chromosome)))
g$chromosome=as.character(g$chromosome)
names(colgene)=unique(g$chromosome)
####################

 breaks <- c(seq(0.7, 0.99, length.out = 101),seq(1,1.3, length.out = 101))
 #breaks <- c(seq(0, 1, length.out = 101),seq(1.01,2, length.out = 101))

k <- c("#818F72", "#C7DEC1", "#A0D439", "#B0F24E")
names(k)= as.character(seq_len(4))
wtlcol=list( colgene,  k)
names(wtlcol)= c("chromosome","kp")
###Plot the heatmap:
stopifnot(colnames(p16)== rownames(pc))

#stopifnot(colnames(p16)== rownames(md))

  png(file = paste0(hDir, "/HEATMAP_P16_Cnv_Bright_3.png") ,height=1300,width=1300,res=300)
 pheatmap::pheatmap(p16, show_rownames = F, show_colnames = F,   annotation_col = pc, cluster_rows = F, cluster_cols = T,annotation_row = g, color = colorRampPalette(c("navy", "white", "firebrick3"))(200),annotation_colors= wtlcol,clustering_method =  "ward.D",clustering_distance_rows="correlation", annotation_names_row=T, breaks = breaks) 
dev.off()
```


LP: No consensus cluster found
```{r}
lp= obs1[,which(colnames(obs1) %in% rownames(cnt@meta.data)[which(cnt$subtype=="LP")]  )]
md <- epith@meta.data[which(colnames(epith) %in% colnames(lp)), c(1,21)] %>% as.data.frame()
md=md[colnames(lp),]
n <- rownames(md)
md=as.data.frame(md[,2], row.names = n)
colnames(md)= "subtype"
md$subtype=as.character(md$subtype)
subtype="#7D726A"
names(subtype)="LP"
wtlcol=list(colgene, subtype)
names(wtlcol)= c("chromosome","subtype")
##
  png(file = paste0(hDir, "/HEATMAP_LP_Cnv_Bright_5.png") ,height=1500,width=1500,res=300)
 pheatmap::pheatmap(lp, show_rownames = F, show_colnames = F,   annotation_col = md, cluster_rows = F, cluster_cols = T,annotation_row = g, color = colorRampPalette(c("navy", "white", "firebrick3"))(200),annotation_colors= wtlcol,clustering_method =  "ward.D",clustering_distance_rows="correlation", annotation_names_row=T, breaks = breaks) 
dev.off()
```




