---
title: "Step1_1_Preprocessing"
author: "Melissa"
date: "6/22/2021"
output: html_document
---

```{r setup, include=FALSE, message=F}
knitr::opts_chunk$set(echo = F)
```

Load the global variables (mainly the libraries and created folders to store results)
```{r}
mainDir <- here::here()
source(knitr::purl(file.path(mainDir,"Scripts","global_var_current.Rmd"), quiet=TRUE))
source(knitr::purl(file.path(mainDir,"Scripts","scRNAseq_Functions.Rmd"), quiet=TRUE))
source(knitr::purl(file.path(mainDir,"Scripts","global_variables.Rmd"), quiet=TRUE))
```

#---------------------- Automatic Creation of Seurat Objects with Metadata -----------------------
--------------------------------------------------------------------------------------------------

##Create Seurat Objects from CellRanger exported expression matrices:
```{r load expression matrices as separate seurat objects}
## Make sure the input matrices are stored in sub directories inside each sample name folder :

ld <- list.dirs(inputDir)[seq(4,14,2)] #keep only paths to the expressions matrices
samplename <- gsub(paste0(inputDir,"/invivo/"), "", list.dirs(inputDir)[seq(3,13,2)])

for (i in seq (1, length(ld))) {
  message(paste0("Creating Seurat Object for Sample: ", samplename[i]))
  assign(samplename[i], CreateSeuratObject(Read10X(ld[i]), project = samplename[i], min.cells = 10, min.features = 50)) #keep genes detected in at least 10 cells, and cells having at least 50 detected genes
}
gc()
```

Calculate % of mitochondrial genes and plot basic QC representations to estimate the median values of number of genes; UMIs detected overall and depict outliers:


```{r Add Metadata}
seuratlist= mget(samplename)
metadata= c("Ctrl_2.7M", "Ctrl_3M", "Ctrl_5.2M", "T6413", "T94", "T99")
```


```{r}
for(i in seq(1:length(seuratlist))){
  seuratlist[[i]][["percent.mito"]] <- PercentageFeatureSet(seuratlist[[i]], pattern = "^mt-")
}
features= c("nCount_RNA", "nFeature_RNA", "percent.mito")
pdf(paste0(QCdir, "/VlnPlot_Samples_BeforeQC.pdf") 
   , height = 5, width = 5)
 for(i in seq(1,5)){
p=initialfeaturesPlot(seuratlist[[i]], qcfeatures= features)
print(p)
 }
 dev.off()
```

```{r Add Metadata}
## merge the individual samples into a single object by keeping the metadata related to each sample:
brca <- merge(x=seuratlist[[1]], y=seuratlist[c(2: length(seuratlist))], add.cell.ids = metadata)
brca[["percent.mito"]] <- PercentageFeatureSet(brca, pattern = "^mt-")
```



#------------------ Set QC filters on the merged samples & Perform SCT Normalization, Dim Reduction & Clustering ------------------------#
------------------------------------------------------------------------------------------------------------------------------------------
```{r}
brca <- brca%>% subset(., nFeature_RNA < 10000 & nFeature_RNA > 1000 
                            & nCount_RNA < 100000  
                                  & percent.mito < 30) %>% SCTransform(., method = "glmGamPoi",  verbose = FALSE)    %>% RunPCA(., npcs = 60)  %>% RunUMAP(., reduction= "pca", dims= 1:60 ) %>% FindNeighbors(., reduction = "pca", dims = 1:60) %>% FindClusters(., resolution = 1.2)
```

Clean the working directory:
If you are running this step on Windows, the names are changed: the "_" after Cre is "+"
```{r}
rm(T6413)
rm(T99)
rm(T94)
rm(Control_Cre_2.7M)
rm(Control_Cre_5.2M)
rm(Control_Cre_3M)
gc()
```


```{r}
mt= brca@meta.data[,c("nCount_RNA","nFeature_RNA","percent.mito", "orig.ident")]
 pdf(file = paste0(QCdir, "/nUMIvsnGene.pdf"),   width = 15, height = 20)
  ggplot(mt, aes(x=nCount_RNA, y=nFeature_RNA, col= percent.mito)) + geom_point(alpha = 0.7) + scale_color_viridis(option="B") +  xlab("nGenes") + ylab("nUMI") +  labs(fill = "% MT genes") +facet_wrap(~ orig.ident, ncol = 2) +gtheme
dev.off()
rm(mt)
```

#------------------------- Evaluate clustering resolution ---------------------------------------#
----------------------------------------------------------------------------------------------------------------------------


Check the phylogenetic tree over the cell identities present in each Seurat_cluster:
```{r}
#Create an object for that:
seu= brca
DefaultAssay(seu)="SCT"
# Select a range of resolutions
resolution.range <- seq(from = 0.6, to = 1.2, by = 0.2)
# Find clusters using a range of resolutions
seu <- Seurat::FindClusters(object = seu, resolution = resolution.range)

Idents(seu)="seurat_clusters"
seu <-BuildClusterTree(seu, assay = "SCT")
## plot the tree & save it :
pdf(file = paste0(QCdir, "/ResolutionClustering_1k10K_SCT.pdf"),  
    width = 15, 
    height = 12)
p <- clustree::clustree(seu)
print(p)
dev.off()
rm(seu)
gc()
```

#--------------Create Additional metadata ------------------------------------------------------#
------------------------------------------------------------------------------------------------------------------------------------

##Histological : No lesions, Precanc lesions; tumor 1; tumor2; tumor3
##Tumor size
##celltype
#subtype according to gene expression
```{r}
Idents(brca)= "orig.ident"
##
brca$sample=brca$orig.ident
brca$histology= brca$orig.ident
brca$histology= recode(brca$histology,"Control_Cre_2.7M" = "WT",
                       "Control_Cre_3M"   = "WT",
                       "Control_Cre_5.2M" = "WT_PrimaryLesions",
                       "T6413"="Tumor",
                       "T94"  ="Tumor",
                       "T99" ="Tumor" )
##
tumorsize= c("WT","WT","WT_PrimaryLesions","LargeTumor","MediumTumor","SmallTumor")
names(tumorsize)= levels(Idents(brca))
brca=RenameIdents(brca, tumorsize)
brca$tumorsize= Idents(brca)
```


#--------------Estimate Cell Cycle Phase ------------------------------------------------------#
------------------------------------------------------------------------------------------------


```{r}
#Load the S & G2M genes:
s.genes <- Seurat::cc.genes$s.genes %>% tolower() %>% Hmisc::capitalize()
g2m.genes <- Seurat::cc.genes$g2m.genes %>% tolower() %>% Hmisc::capitalize()
brca= CellCycleScoring(brca, s.features = s.genes, g2m.features = g2m.genes, 
                                        set.ident = FALSE, nbin= 30)
rm(s.genes); rm(g2m.genes)
```


save the object:
```{r}
qs::qsave(brca, paste0(RDatadir,  "/BRCA_SCT_1k10k.qs"))
```

