---
title: "Step4_TF_Enrichment"
author: "Melissa"
date: "18/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
mainDir= mainDir= here::here()
inputDir= file.path(mainDir, "input", "human", "snRNAseq", "juxtaTumor","filtered_matrices")
outputDir= file.path(mainDir, "output", "human","snRNAseq", "juxtaTumor")
#####################
RDatadir <- file.path(outputDir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}
QCdir <- file.path(outputDir, "QC") ; if(!file.exists(QCdir)){dir.create(QCdir)}
DEdir <- file.path(outputDir, "DiffExp") ; if(!file.exists(DEdir)){dir.create(DEdir)}
clustDir= file.path(outputDir, "Clustering") ; if(!file.exists(clustDir)){dir.create(clustDir)}
cnvDir= file.path(outputDir, "CNV") ; if(!file.exists(cnvDir)){dir.create(cnvDir)}
TFDir= file.path(outputDir, "TF_Enrichment") ; if(!file.exists(TFDir)){dir.create(TFDir)}
scenicDir= file.path(TFDir, "SCENIC") ; if(!file.exists(scenicDir)){dir.create(scenicDir)}
cheaDir= file.path(TFDir, "CHEA3") ; if(!file.exists(cheaDir)){dir.create(cheaDir)}
```

RUN CHEA3:
```{r}

library(httr)
library(jsonlite)
##COMPARISON WITH LP:
genes = comp$gene

url = "https://maayanlab.cloud/chea3/api/enrich/"
encode = "json"
payload = list(query_name = "myQuery", gene_set = genes)
	#POST to ChEA3 server
response = POST(url = url, body = payload, encode = encode)
json = httr::content(response, as="text")
	#results as list of R dataframes
results = fromJSON(json)
dt= results[["Integrated--meanRank"]]
dt$Score=as.numeric(as.character(dt$Score))
dt$Rank=as.numeric(as.character(dt$Rank))

writexl::write_xlsx(dt, paste0(cheaDir, "/Chea3_OverexpressedGenes_LP2_vsLP1_FC1.xlsx"))
dt= readxl::read_excel(paste0(cheaDir, "/Chea3_OverexpressedGenes_LP2_vsLP1_FC1.xlsx"))
```

Work on the score:
```{r}
library(ggpubr)
library(ggrepel)
dt$Top=ifelse(dt$Score <= 100, "Top", "No")

pdf(paste0(cheaDir, "/Fig1_Plot_Rank_logScore_LP1vsLP2_log1.pdf"))
p <- dt %>% ggplot(aes(y= log(Score), x=Rank)) + geom_point() + theme_bw() + geom_hline(yintercept = 4, color= "red") 
print(p)
p <- dt %>% ggplot(aes(x= log(Score))) + geom_density() + theme_bw()
print(p)

p <- dt %>% ggplot(aes(y= log(Score), x=Rank)) + theme_bw() + geom_hline(yintercept = 4.2, color= "red")+
  geom_text_repel(
    data = subset(dt, Score <= 100),
    aes(label = TF),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")) + geom_point(aes(color = Top)) +
  scale_color_manual(values = c( "grey","red"))
print(p)
dev.off()
```

Only for Top TFs:
```{r}
top_tf=dt$TF[1:20] 
top_TF=dt[1:20,]
#####
tf.expression= PrctCellExpringGene(epithelial, genes= top_tf, group.by = "subtype" )
tf.expressiona= PrctCellExpringGene(epithelial, genes= top_tf, group.by = "all" )
tf.expressiona$rank= seq(1, nrow(tf.expressiona))
df= tf.expression %>% dplyr::filter(Feature== "LP_2")

tf.expressiona$isexpP16= ifelse(df$Cell_proportion > 0.2, "High_pct_Expression", "Low_pct_Expression")

l=list("Percentage_Exp_Subtype"=tf.expression , "Percentage_Exp_All"=tf.expressiona, "OnlyPretumoralCells"= dt)
writexl::write_xlsx(l,paste0(cheaDir, "/Chea3_TopTF_LP_2_Juxta_log1.xlsx"))
```

```{r}
library(viridis)
pdf(paste0(cheaDir, "/Top30_TFs_Chea3_Juxta_LP2_log1.pdf"))
for(i in seq_along(top_tf)[-6]){ #length(top_tf)
  p <- FeaturePlot(epithelial, features = top_tf[i], pt.size = 1 ) + NoAxes()  + scale_color_viridis(option="B",limits = c(0, 2.5), na.value= "grey",direction=-1)
p[[1]]$layers[[1]]$aes_params$alpha = 0.4
print(p)
}
dev.off()
```


##---------- CREATE A GLOBAL DATASET-----------##
```{r}
df= tf.expression %>% dplyr::filter(Feature== "LP_2")
df$TF= toupper(df$Markers)
top_TF= merge(top_TF, df, by="TF")

top_TF$Cell_proportion= ifelse(is.na(top_TF$Cell_proportion)==TRUE, 0, top_TF$Cell_proportion)
###
top_TF$isexpP16= ifelse(top_TF$Cell_proportion >= 0.1, "Expressed", "NotExpressed")

top_TF$isexpP16= ifelse(is.na(top_TF$isexpP16)==TRUE, "NotExpressed", top_TF$isexpP16)



writexl::write_xlsx(top_TF, paste0(cheaDir, "/Top_TF_PretumoralCells_ExpressionPct_LP2cells_log1.xlsx"))
top_TF= readxl::read_excel(paste0(cheaDir, "/Top_TF_PretumoralCells_ExpressionPct_LP2cells.xlsx"))
```


Plot the score, Rank and percentage of expression:
```{r}
pdf(paste0(cheaDir, "/DotPlot_Top30Genes_CellProp_Rank_Score_LP2_Juxta_log1.pdf"))
p <- top_TF %>% ggplot(aes(y= log(Score), x=Rank, size= Cell_proportion, color= Cell_proportion)) + theme_bw() + geom_hline(yintercept = 4.2, color= "red")+
  geom_text_repel(
    data = subset(top_TF, Score <= 100),
    aes(label = TF),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")) + geom_point() +
  scale_color_viridis(option="B")
print(p)
dev.off()
```


SCENIC:

```{r}
setwd(scenicDir)

dir.create("SCENIC/cisTarget_databases") # Build the database folder
#setwd("C:/Users/Melissa S/Desktop/BRCA1_Tumorigenesis/Organoid_Output/SCENIC/cisTarget_databases")

setwd(paste0(scenicDir,"/cisTarget_databases"))

dbFiles <- c("https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-500bp-upstream-7species.mc9nr.genes_vs_motifs.rankings.feather",
"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather")

for(featherURL in dbFiles)
{
  download.file(featherURL, destfile=basename(featherURL)) # saved in current dir
}

dbPath <-  "SCENIC" # Choose the appropriate database/location
SCENIC::dbLoadingAttempt(dbPath)

rnk <- RcisTarget::importRankings("C:/Users/Melissa S/Desktop/BRCA1_Tumorigenesis/Juxta_Frozen_Samples_All/script/SCENIC", indexCol="features", columns=c("SOX10")) # you can load only a few genes to do tests faster
rnk
getwd()

###########
setwd(scenicDir)

dbFiles <- c("https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-500bp-upstream-7species.mc9nr.feather",
             "https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather")
## mc9nr: Motif collection version 9: 24k motifs

## File names after download
downloadedDBFiles <- c("cisTarget_databases/hg19-500bp-upstream-7species.mc9nr.feather",
                       "cisTarget_databases/hg19-tss-centered-10kb-7species.mc9nr.feather")


## The databases are downloaded only in the case when the files do not exist.
if( !(file.exists(downloadedDBFiles[1]) & file.exists(downloadedDBFiles[2])) ){
  
  ## Create and move to the new folder
  dir.create("cisTarget_databases");
  setwd("cisTarget_databases") 
  for(featherURL in dbFiles)
  {
    download.file(featherURL, destfile=basename(featherURL)) # saved in current dir
    descrURL <- gsub(".feather$", ".descr", featherURL)
    if(file.exists(descrURL)) download.file(descrURL, destfile=basename(descrURL))
  }
  setwd("..")
  
}

rnk <- RcisTarget::importRankings(scenicDir, indexCol="features", columns=c("SOX10")) # you can load only a few genes to do tests faster
rnk

library(SCENIC)
org <- "hgnc"
dbDir <- setwd(paste0(scenicDir,"/cisTarget_databases")) # RcisTarget databases location

myDatasetTitle <- "SCENIC test" # choose a name for your analysis
data(defaultDbNames)
dbs <- defaultDbNames[[org]]


scenicOptions <- initializeScenic(org=org, dbDir=dbDir, dbs=dbs, datasetTitle=myDatasetTitle, nCores=1)  

scenicOptions=readRDS()
epithelial2=epithelial
epithelial= subset(epithelial, subtype %in% c("LP_1", "LP_2", "Basal"))

cellInfo <- data.frame(epithelial@meta.data)
colnames(cellInfo)[which(colnames(cellInfo)=="orig.ident")] <- "sample"
colnames(cellInfo)[which(colnames(cellInfo)=="nFeature_RNA")] <- "nGene"
colnames(cellInfo)[which(colnames(cellInfo)=="nCount_RNA")] <- "nUMI"
colnames(cellInfo)[which(colnames(cellInfo)=="seurat_clusters")] <- "cluster"
colnames(cellInfo)[which(colnames(cellInfo)=="subtype")] <- "celltype"
cellInfo <- cellInfo[,c("sample","nGene","nUMI","cluster","celltype")]
#----------------------------------###
scenicOptions@inputDatasetInfo$cellInfo <- cellInfo
scenicOptions@inputDatasetInfo$colVars <- as.list(cols_ep)

saveRDS(scenicOptions, file=paste0(scenicDir,"/int_2/scenicOptions.Rds"))


scenicOptions@settings$modules$weightThreshold= 0.05



exprMat <- as.matrix(epithelial@assays$RNA@counts) #Prepare the expression matrix. In order to save computing resources, only some cells are randomly selected to calculate the co-expression network

genesKept <- geneFiltering(exprMat, 
                           scenicOptions=scenicOptions,minCountsPerGene=10*.01*ncol(exprMat),minSamples=ncol(exprMat)*.1) #Gene filtering/selection, remove genes that are most likely to be noise

genesKept= readRDS(paste0(scenicDir,"/int/1.1_genesKept.Rds"))

exprMat_filtered <- exprMat[genesKept, ]

dim(exprMat_filtered)

runCorrelation(exprMat_filtered, scenicOptions) ##Calculate the correlation matrix, 1.2_corrMat.Rds: the correlation matrix between genes

##GENIE3, the input of GENIE3 is usually an expression matrix and a list of candidate regulators.
exprMat_filtered <- log2(exprMat_filtered+1) #complete matrix

runGenie3(exprMat_filtered, scenicOptions, nParts = 10,resumePreviousRun = TRUE) #nParts parameter, is to divide the expression matrix into n parts and calculate separately
#1.4, GENIE3_linkList.Rds: The final result of GENIE3 is to merge the files starting with "1.3_"

library(doParallel)
library(doRNG)
getwd()
exprMat_log <- log2(exprMat+1) #log standard words original matrix

scenicOptions@settings$verbose <- TRUE
scenicOptions@settings$nCores <- 1

scenicOptions@settings$seed <- 123
scenicOptions@settings$modules$weightThreshold= 0.001


res <- runSCENIC_1_coexNetwork2modules(scenicOptions=scenicOptions,corrThr = 0.1, corrMat = corrMat, linkList = linklist) #1. Get coexpression module


scenicOptions <- runSCENIC_2_createRegulons(scenicOptions,coexMethod=c("top10perTarget"))
scenicOptions <- runSCENIC_3_scoreCells(scenicOptions, exprMat_log) #3. Scoring the GRN (regulator) in the cell
saveRDS(scenicOptions, file="int/scenicOptions.Rds")
runSCENIC_4_aucell_binarize(scenicOptions, exprMat=exprMat_log)
```





```{r}
##Import the original regulonAUC matrix, which is the AUC matrix generated after running runSCENIC_3_scoreCells, you can view the AUC activity score of each GRN in each cell

AUCmatrix <- readRDS("int/3.4_regulonAUC.Rds")
AUCmatrix <- data.frame(t(AUCmatrix@assays@data@listData$AUC), check.names=F)
RegulonName_AUC <- colnames(AUCmatrix)
RegulonName_AUC <- gsub(' \\(','_',RegulonName_AUC) #replace "(" with "_"
RegulonName_AUC <- gsub('\\)','',RegulonName_AUC) # Remove ")" and finally replace KLF3_extended (79g) with KLF3_extended_79g
colnames(AUCmatrix) <- RegulonName_AUC
library(Seurat)
sseu=epithelial
sseu <- AddMetaData(sseu, AUCmatrix) # Add the AUC matrix to the metadata information of pbmc
```


Plots of TFs:
Visualize the top genes:
```{r}
library("gplots")
library(ComplexHeatmap)
suppressPackageStartupMessages(library(ComplexHeatmap))


metadata= data.frame("subtype"=sseu@meta.data$subtype)
rownames(metadata)= colnames(sseu)
#metadata= metadata[, -1] %>% as.data.frame()
colours <- list("subtype"=cols_ep)
####---------------------##
colAnn <- HeatmapAnnotation(df = metadata,
  which = 'col',
  col = colours,
  annotation_width = unit(c(1, 4), 'cm'),
  gap = unit(0.5, 'mm'))
identical(rownames(metadata), rownames(AUCmatrix))
```

```{r}
dt= sseu@meta.data
dt= dt %>% dplyr::select(grep("g$", colnames(dt), value = TRUE))
dt= dt %>% select(! grep("extended",colnames(dt), value = TRUE))
```


```{r}
split <- factor( metadata$subtype,levels = c("Basal","LP_1","LP_2", "Avd","ML" ))

hmap <- Heatmap(
  t(AUCmatrix),
  name = "TF SCENIC module scores",
 # col = bluered(75), 
  show_row_names = TRUE,
  show_column_names = FALSE,
  cluster_rows =TRUE,
  cluster_columns = FALSE,
  show_column_dend = FALSE,
  show_row_dend = TRUE,
  row_dend_reorder = FALSE,
  column_dend_reorder = FALSE,
  clustering_method_rows = "ward.D2",
  clustering_method_columns = NULL,
  #width = unit(100, "mm"),
  top_annotation=colAnn,row_names_gp = gpar(fontsize = 8, fontfamily = "sans"),column_split=split,column_names_gp = gpar(fontsize = 4, fontfamily = "sans"))

pdf(paste0(scenicDir, "Heatmap_SCENIC_AUCMatrix_ScoreModules_SeuSubset.pdf"), height = 10, width = 12)
draw(hmap, heatmap_legend_side="right", annotation_legend_side="right")
dev.off()
#---------
png(paste0(scenicDir, "Heatmap_SCENIC_AUCMatrix_ScoreModules_JUXTAPart1.png"), height = 2000, width = 2800, res=300)
print(hmap)
dev.off()
#draw(hmap, heatmap_legend_side="right", annotation_legend_side="right")
```